connection/server.py

```diff
@@ -0,0 +1,216 @@
+import pickle
+import socket
+import sys
+import time
+import threading
+import numpy as np
+
+from game_simulation.game import Game
+
+NEW_CONNECTION = 1
+
+CONNECTION_IS_SUCCESSFUL = b"OK"
+SERVER_IS_FULL = b"NO"
+
+SERVER_IS_CLOSING = True
+SERVER_IS_RUNNING = False
+
+CLOSED_CONNECTION = 1
+QUIT = "q"
+CLOSE_SIGNAL = None
+
+GAME_IS_ENDED = 1
+ROUND_TIME = 1
+MAX_PLAYERS = 3
+SEMAPHORE_INIT_VALUE = 0
+NEW_GAME = 1
+
+SERVER_IP = "127.0.0.1"
+SERVER_PORT = 6000
+SIZE_OF_CLIENT_MESSAGE = 34
+
+NUMBER_OF_THREADS_TO_CANCEL_CONNECTION = 1
+NUMBER_OF_STARTED_CONNECTIONS_INIT_VALUE = 0
+NUMBER_OF_STARTED_GAMES_INIT_VALUE = 0
+NOT_NEED_OF_NEW_CONNECTIONS = 0
+
+CLIENT_IS_CONNECT = False
+CLIENT_IS_CLOSING = True
+
+FREE = True
+BUSY = False
+
+GAME_NOT_TO_START = False
+GAME_TO_START = True
+
+EMPTY_PLACE_FOR_THREADS = np.array([None, None])
+EMPTY_PLACE_FOR_SOCKET = None
+
+
+class Server:
+
+    def __init__(self):
+        # server's variables
+        self.ip = SERVER_IP
+        self.port = SERVER_PORT
+        self.server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.server_socket.bind((self.ip, self.port))
+        self.server_socket.listen(MAX_PLAYERS)
+        self.server_status = SERVER_IS_RUNNING
+
+        # clients' variables
+        self.client_sockets = np.array([EMPTY_PLACE_FOR_SOCKET for _ in range(MAX_PLAYERS)])
+        self.client_status = np.array([CLIENT_IS_CONNECT for _ in range(MAX_PLAYERS)])
+        self.client_threads = np.array([EMPTY_PLACE_FOR_THREADS for _ in range(MAX_PLAYERS)])
+
+        # variables for connections management
+        self.number_of_started_connections = NUMBER_OF_STARTED_CONNECTIONS_INIT_VALUE
+        self.connections_needed_to_start = NOT_NEED_OF_NEW_CONNECTIONS
+        self.number_of_started_games = NUMBER_OF_STARTED_GAMES_INIT_VALUE
+        self.client_number = np.array([FREE for _ in range(MAX_PLAYERS)])
+        self.game_to_start = np.array([GAME_NOT_TO_START for _ in range(MAX_PLAYERS)])
+
+        # variables for game
+        self.game = Game()
+        self.game_lock = threading.Lock()
+        self.start_round = [threading.Semaphore(SEMAPHORE_INIT_VALUE) for _ in range(MAX_PLAYERS)]
+
+    def start_server(self):
+        threading.Thread(target=self.open_server_commands).start()
+
+        for x in range(MAX_PLAYERS + NUMBER_OF_THREADS_TO_CANCEL_CONNECTION):
+            threading.Thread(target=self.connect_client).start()
+
+        threading.Thread(target=self.round_timer).start()
+
+        while True:
+            self.start_games()
+            self.open_new_connections()
+
+            if self.server_status == SERVER_IS_CLOSING:
+                break
+
+        # wait for clients disconnection for closing server
+        for client_threads in self.client_threads:
+            for thread in client_threads:
+                while thread.is_alive():
+                    continue
```

Comment:
When I try to close the server by 'q' into its console, it works just fine (client are getting notified about that), but server throws this ugly error, can this be somehow prevented?

![obraz](https://github.com/Misiek22113/PIO_CROSSY_ROAD/assets/71559210/cf3a73fe-9ea5-465e-91b7-83179d5c99eb)

Reply:
When you force close, it is KeyboardInterrupt. I think we don't need to handle that